src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js

```diff
@@ -0,0 +1,683 @@
+   * keyboard shortcut or button event.
+   */
+  initBrushing(selectedZoomBtn, zoom) {
+    let extent;
+    const brushClass = "brush";
+
+    const brush = d3
+      .brushX()
+      .extent([
+        [0, 0],
+        [this.canvas.inner.width, this.canvas.inner.height],
+      ])
+      .on("start " + brushClass, updateChart);
+
+    // The brush element must be child of zoom element - this is only way we found to have both
+    // zoom and brush events working at the same time. See https://stackoverflow.com/a/59757276 .
+    const brushElem = zoom.zoomElem
+      .append("g")
+      .attr("class", brushClass)
+      .call(brush);
+
+    const self = this;
+
+    /**
+     * Zooms into selected fragment of plot.
+     *
+     * Based on https://www.d3-graph-gallery.com/graph/interactivity_brush.html
+     * Section "Brushing for zooming".
+     */
+    const zoomIn = () => {
+      const xMin = self.scale.x.invert(extent[0]);
+      const xMax = self.scale.x.invert(extent[1]);
+
+      self.scale.x.domain([xMin, xMax]);
+      const dx = extent[1] - extent[0];
+      self.scale.zoom = self.scale.zoom * (self.canvas.inner.width / dx);
+
+      self.rescale(self.scale, true);
+    };
+
+    const zoomInKeyEvent = (event) => {
+      if (shortcuts.zoomIn(event)) {
+        zoomIn();
+        endBrushing();
+      }
+    };
+
+    /**
+     * Updates plot when brushing.
+     */
+    function updateChart() {
+      const selectionEvent = d3.event.selection;
+      selectedZoomBtn.style.display = "inline-block";
+      selectedZoomBtn.addEventListener("click", zoomIn, true);
+      document.addEventListener("keydown", zoomInKeyEvent, true);
+      extent = selectionEvent;
+    }
+
+    /**
+     * Removes brush, keyboard event and zoom button when end event is captured.
+     */
+    const endBrushing = () => {
+      brushElem.call(brush.move, null);
+      selectedZoomBtn.style.display = "none";
+      selectedZoomBtn.removeEventListener("click", zoomIn, true);
+      document.removeEventListener("keydown", zoomInKeyEvent, true);
+    };
+
+    let endEvents = ["click", "auxclick", "contextmenu", "scroll"];
+    endEvents.forEach((e) => document.addEventListener(e, endBrushing, false));
+  }
+
+  /**
+   * Helper function for rescaling the data points with a new scale.
+   */
+  rescale(scale, with_animation) {
+    const animation_duration = with_animation ? ANIMATION_DURATION : 0.0;
+    this.xAxis
+      .transition()
+      .duration(animation_duration)
+      .call(d3.axisBottom(scale.x).ticks(this.binCount()));
+    this.plot
+      .selectAll("rect")
+      .transition()
+      .duration(animation_duration)
+      .attr(
+        "transform",
+        (d) =>
+          "translate(" +
+          scale.x(d.x0) +
+          "," +
+          scale.y(d.length) +
+          ")scale(" +
+          scale.zoom +
+          ",1)"
+      );
+  }
+
+  /**
+   * Update the d3 histogram with the current data.
```

Comment:
Write few momre words how the update is performed pls. Is all graphical data thrown away and re-drawn, or is it done differently?

Reply:
Extended the description. This has been reworked to no longer re-do the whole vis, but to do a proper update,